### PR TITLE
CDAP-1663: Refactor Id.Program to include ProgramType

### PR DESCRIPTION
**Pull request by jwang47**
First part of several large refactoring PRs to enable a clean implementation of application authorization.

### CODE REVIEW

File: cdap-app-fabric/src/main/java/co/cask/cdap/app/program/ManifestFields.java [hunk 1]
```diff
@@ -25,7 +25,7 @@
 public final class ManifestFields {
   public static final Attributes.Name MAIN_CLASS = Attributes.Name.MAIN_CLASS;
   public static final Attributes.Name MANIFEST_VERSION = Attributes.Name.MANIFEST_VERSION;
-  public static final Attributes.Name PROCESSOR_TYPE = new Attributes.Name("Processor-Type");
+  public static final Attributes.Name PROGRAM_TYPE = new Attributes.Name("Program-Type");
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/app/program/ManifestFields.java]: So this will break jars that are already deployed?

Comment (jwang47) [inline cdap-app-fabric/src/main/java/co/cask/cdap/app/program/ManifestFields.java]: Will revert this change.

**Comment by jwang47** [comment]
Build passed http://builds.cask.co/browse/CDAP-DUT941-5

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java [hunk 2]
```diff
@@ -955,8 +957,8 @@ public void liveInfo(HttpRequest request, HttpResponder responder, @PathParam("n
                        @PathParam("program-id") String programId) {
     ProgramType type = getProgramType(programCategory);
     if (type == null) {
-      responder.sendString(HttpResponseStatus.METHOD_NOT_ALLOWED, String.format("Live-info not supported for program" +
-                                                                                  " type '%s'", programCategory));
+      responder.sendString(HttpResponseStatus.BAD_REQUEST,
```
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java]: Should we send a review for this response code change?

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/util/AbstractAppFabricHttpHandler.java [hunk 3]
```diff
@@ -332,18 +332,27 @@ protected boolean respondIfElementNotFound(Throwable t, HttpResponder responder)
...
+                                Data type, String namespaceId, String name, String appId) {
     try {
       if ((name != null && name.isEmpty()) || (appId != null && appId.isEmpty())) {
         responder.sendString(HttpResponseStatus.BAD_REQUEST, "Empty name provided");
         return;
       }
 
-      Id.Program program = Id.Program.from(namespace, appId == null ? "" : appId, "");
-      String json = name != null ? getDataEntity(store, dsFramework, program, type, name) :
-        appId != null ? listDataEntitiesByApp(store, dsFramework, program, type)
```
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/util/AbstractAppFabricHttpHandler.java]: huh? was there a bug here that you fixed?

Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/util/AbstractAppFabricHttpHandler.java]: looks like you did the right thing, just wondering how we didn't catch this bug so far. am I missing something?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/DistributedProgramRuntimeService.java [hunk 4]
```diff
@@ -200,7 +200,7 @@ public synchronized RuntimeInfo lookup(final RunId runId) {
           continue;
         }
 
-        Id.Program programId = Id.Program.from(matcher.group(2), matcher.group(3), matcher.group(4));
+        Id.Program programId = Id.Program.from(matcher.group(2), matcher.group(3), appType, matcher.group(4));
```
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/DistributedProgramRuntimeService.java]: appType?

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/util/AbstractAppFabricHttpHandler.java (hunk 3 above)
Comment (jwang47) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/util/AbstractAppFabricHttpHandler.java]: I don't think I fixed any bugs here, just refactored it to be easier to read and also not pass unused information. For example `listDataEntitiesByApp(store, dsFramework, program, type)` doesn't need `Id.Program program`, it only needs `Id.Application`.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/DistributedProgramRuntimeService.java (hunk 4 above)
Comment (jwang47) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/DistributedProgramRuntimeService.java]: Not sure why the local variable was previously named `appType`. Anyway, the local variable is not needed, so I can remove it.

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java (hunk 2 above)
Comment (jwang47) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java]: I think it's straightforward, but I probably used the wrong HTTP code here.

`Method Not Allowed` is for when you make a request to a path that doesn't support the HTTP method in your request. In this case, `Not Found` would probably make the most sense since conceptually there isn't a live-info endpoint for program types that don't support it.

Comment (jwang47) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java]: But actually this isn't really in the scope of this PR, so I'll revert that change.

File: cdap-proto/src/main/java/co/cask/cdap/proto/ProgramType.java [hunk 5]
```diff
@@ -61,28 +106,41 @@
...
-    return categoryName;
+    return parameters.getCategoryName();
   }
 
   public String getPrettyName() {
-    return prettyName;
+    return parameters.getPrettyName();
+  }
+
+  public SchedulableProgramType getSchedulableType() {
```
Comment (bdmogal) [inline cdap-proto/src/main/java/co/cask/cdap/proto/ProgramType.java]: ``@Nullable``?

File: cdap-proto/src/main/java/co/cask/cdap/proto/ProgramType.java [hunk 6]
```diff
@@ -99,7 +157,82 @@ public static ProgramType valueOfCategoryName(String categoryName) {
...
+      Preconditions.checkArgument(prettyName != null, "prettyName cannot be null");
+      Preconditions.checkArgument(listable != null, "listable cannot be null");
+      Preconditions.checkArgument(categoryName != null, "categoryName cannot be null");
+      this.prettyName = prettyName;
+      this.listable = listable;
+      this.categoryName = categoryName;
+      this.schedulableType = schedulableType;
+    }
+
+    public SchedulableProgramType getSchedulableType() {
```
Comment (bdmogal) [inline cdap-proto/src/main/java/co/cask/cdap/proto/ProgramType.java]: ``@Nullable`` here too? Although where is this used?

**Comment by bdmogal** [comment]
Overall, just minor comments. However, I am not too sure about the manifest/jar part of the code that this PR modifies. Has @chtyim reviewed it already?

**Comment by jwang47** [comment]
Yes, @chtyim did. I reverted the change to the manifest field name, so it's only changing the variable name now.

**Comment by bdmogal** [comment]
Ok. The other stuff LGTM.

**Comment by bdmogal** [comment]
Apart from the two minor ``@Nullable`` comments :)

**Comment by jwang47** [comment]
Build passed http://builds.cask.co/browse/CDAP-DUT941-7
